plutus-core/testlib/PlutusCore/Generators/PIR.hs

```diff
@@ -0,0 +1,1752 @@
+                                  }
+
+-- | Bind type names
+bindTyNames :: [(TyName, Kind ())] -> GenTm a -> GenTm a
+bindTyNames = flip $ foldr (uncurry bindTyName)
+
+-- | Remember that we have generated a type name locally but don't bind it.
+-- Useful for non-recursive definitions where we want to control name overlap.
+registerTyName :: TyName -> GenTm a -> GenTm a
+registerTyName n = local $ \ e -> e { geUnboundUsedTyNames = Set.insert n (geUnboundUsedTyNames e) }
+
+-- | Bind a term to a type in a generator.
+bindTmName :: Name -> Type TyName DefaultUni () -> GenTm a -> GenTm a
+bindTmName x ty = local $ \ e -> e { geTerms = Map.insert x ty (geTerms e) }
+
+-- | Bind term names
+bindTmNames :: [(Name, Type TyName DefaultUni ())] -> GenTm a -> GenTm a
+bindTmNames = flip $ foldr (uncurry bindTmName)
+
+-- | Create a fresh term name, bind it to a type, and use it in a generator.
+bindFreshTmName :: String -> Type TyName DefaultUni () -> (Name -> GenTm a) -> GenTm a
+bindFreshTmName name ty k = do
+  x <- genFreshName name
+  bindTmName x ty (k x)
+
+-- | Get the names and types of the constructors of a datatype.
+constrTypes :: Datatype TyName Name DefaultUni DefaultFun () -> [(Name, Type TyName DefaultUni ())]
+constrTypes (Datatype _ _ xs _ cs) = [ (c, abstr ty) | VarDecl _ c ty <- cs ]
+  where
+    abstr ty = foldr (\ (TyVarDecl _ x k) -> TyForall () x k) ty xs
+
+
+-- | Get the name and type of the match function for a given datatype.
+matchType :: Datatype TyName Name DefaultUni DefaultFun () -> (Name, (Type TyName DefaultUni ()))
+matchType (Datatype _ (TyVarDecl _ a _) xs m cs) = (m, matchType)
+  where
+    fvs = Set.fromList (a : [x | TyVarDecl _ x _ <- xs]) <>
+          mconcat [fvType ty | VarDecl _ _ ty <- cs]
+    pars = [TyVar () x | TyVarDecl _ x _ <- xs]
+    dtyp = foldl (TyApp ()) (TyVar () a) pars
+    matchType = abstr $ dtyp ->> TyForall () r Star (foldr ((->>) . conArg) (TyVar () r) cs)
+      where r = freshenTyName fvs $ TyName $ Name "r" (toEnum 0)
+            conArg (VarDecl _ _ ty) = setTarget ty
+            setTarget (TyFun _ a b) = TyFun () a (setTarget b)
+            setTarget _             = TyVar () r
+    abstr ty = foldr (\ (TyVarDecl _ x k) -> TyForall () x k) ty xs
+
+-- | Bind a datatype declaration in a generator.
+bindDat :: Datatype TyName Name DefaultUni DefaultFun ()
+        -> GenTm a
+        -> GenTm a
+bindDat dat@(Datatype _ (TyVarDecl _ a k) _ _ _) cont =
+  bindTyName a k $
+  local (\ e -> e { geDatas = Map.insert a dat (geDatas e) }) $
+  foldr (uncurry bindTmName) cont (matchType dat : constrTypes dat)
+
+-- | Bind a binding.
+bindBind :: Binding TyName Name DefaultUni DefaultFun ()
+         -> GenTm a
+         -> GenTm a
+bindBind (DatatypeBind _ dat)              = bindDat dat
+bindBind (TermBind _ _ (VarDecl _ x ty) _) = bindTmName x ty
+-- CODE REVIEW: Should we try to generate type bindings and all the recursive types without datatypes stuff?
```

Comment:
The problem with type bindings is that the type-system rules for them are a bit... weird... E.g. you're not allowed to do stuff like (don't remember if this code actually fails, but you get the point):
```
let t : * = int in \ x : t. x
```
because the type synonym `t` would escape the scope - but the type of the function is still ok as `int -> int`... Anyway, I can't remember the details but it turned out to be tricky to manage 😅

Reply:
That particular wart got fixed recently, and type bindings now properly behave like type abstractions.